### ISSUE TITLE
Implement automation in Battle Frontier

### ISSUE THREAD

**Issue by Nox-Ater** [Farigh/pokeclicker-automation#150]
Something simple like entering a loop of starting the Battle Frontier over and over every time you reach the maximum possible.
This to complete Quest and to get Battle Points.

**Comment by linkdest** [comment]
Instead of repeating when reaching the maximum, it would be better reset when you stop one shotting or use a timer with how much seconds (3-30) you want to spend on each floor.

**Comment by Daniferrito** [comment]
As far as I can see, there is no way to reset the battle frontier while getting points. Only letting you fail naturaly grants points.

It would be useful to have the "reset faster" if the objective is simply "kill random mons for egg steps and gems", as the battle frontier gives a decent amount, but the main mode should be to let it finish and restart it then.

**Comment by unproductivityfriendly** [comment]
For the Battle Point Farming Mode: the way of hard reset is to send all the highest attack pokémons to the hatchery queue:

1. Make sure the Setting > Other > Breeding queue is big enough
2. If toggled, remove all the egg step bonuses & PA buff (flutes + oak items + farm)
3. Fill the 4 slots with random pokemons
4. Put pokémon in the hatchery queue by highest attack order
5. After Battle Frontier ends, remove the pokémons from the hatchery queue

**Comment by Farigh** [comment]
It's would not be that simple. The pokemon types is involved as well

**Comment by Daniferrito** [comment]
Calculating the specific breakpoint to when that is possible is not trivial. It depends on the distribution of attack of the pokemons. The more concentrated it is, the earlier it can be done. The more spread out, the harder it is. At some point, with attack completely equal accross the board, putting pokemons into the queue wont make a significant difference.

Other than that, the rewards for the battle frontier scale with the square of the level reached, so it is not that clear how much exiting earlier helps. I will try to make some calculations to see if it actually helps to exit earlier depending in what of the rewards you actually want (money, battle points, egg steps, gems). Money reward is irrelevant, but might as well. Let me know if there is some other incentive to run the battle frontier.

**Comment by Daniferrito** [comment]
After some calculations, here are the results:

![image](https://user-images.githubusercontent.com/2537742/200854388-e89b52d1-11c8-4fa0-9df6-db87ce5ebaec.png)

The attack number at the bottom is the total attack after all multipliers. It varies depending on type matchup, but not by that much. in my case, with 16M pokemon attack, the attack after type moves from 60M to 100M, but mostly stays around 80M.

This graph assumes that you are able to divide your total attack by 5 by putting pokemons into the queue. This is the number I get with my attack distribution, but might be lower or higher for other people.

On the other hand, other than the battle points (because they cannot be gotten anywhere else), the other rewards are almost nothing. With my attack (around 70M):
<details>
  <summary>Efficiency numbers</summary>
  
```Calculating for attack 70,000,000
Max stage: 1271 in 109 minutes
Reduced Max stage: 668 in 20 minutes
Money efficiency: 245.81 per second (1,615,441 every 109 minutes)     
Reduced money efficiency: 371.81 per second (446,224 every 20 minutes)
Money efficiency increase: 151.26%
Battle point efficiency: 2.46 per second (16,154.41 every 109 minutes)
Reduced battle point efficiency: 3.72 per second (4,462.24 every 20 minutes)
Battle point efficiency increase: 151.26%
Egg steps efficiency: 4.60 per second (30,225.97315976563 every 109 minutes)
Reduced egg steps efficiency: 9.60 per second (11,522.666511033787 every 20 minutes)
Egg steps efficiency increase: 208.75%
Gems efficiency: 1.63 per second (10,736 every 109 minutes)
Reduced gems efficiency: 2.61 per second (3,132 every 20 minutes)
Gems efficiency increase: 159.75%
```

From other sources (using the automation):
```Money: 1.3M in 2.7 seconds (481k per second, 129779% of the battle frontier)
Egg steps: 800 steps in 8.5 seconds (94 per second, 1000% of the battle frontier)
Gems: 1000 gems in 16.6 seconds (60 per second, 2300% of the battle frontier)
```
</details>

TL;DR: It is indeed more efficient to leave the battle Frontier earlier, but not by too much. Only farming Battle tokens is worth it, as other sources are WAY better at the other rewards. The efficiency increase by leaving earlier is of 50% (you get 50% more tokens per second).

**Comment by Farigh** [comment]
Let's begin with a simple rerun implementation as a first intention, we'll consider optimisation in a second time

**Comment by linkdest** [comment]
> TL;DR: It is indeed more efficient to leave the battle Frontier earlier, but not by too much. Only farming Battle tokens is worth it, as other sources are WAY better at the other rewards. The efficiency increase by leaving earlier is of 50% (you get 50% more tokens per second).

How is the effiency calculated? Because it doesn't make sense to me. I look this graph and it shows have less attack is more efficient for gems while attack is irrelevent if i dont know what floor i'm.  Let's say I have 1M attack, then I have 145% effiency, how can i translate that to now when stop?

**Comment by Daniferrito** [comment]
> How is the effiency calculated? Because it doesn't make sense to me. I look this graph and it shows have less attack is more efficient for gems while attack is irrelevent if i dont know what floor i'm. Let's say I have 1M attack, then I have 145% effiency, how can i translate that to now when stop?

Ok, I agree the graph is probably quite confusing. It displays the how eficient is to force an early loss compared to letting the run complete normaly, for each level of attack.

So, for example, at 1M attack, you get 50% more battle points per second if you keep forcing resets instead of letting the run finish normaly.

I did not graph how much more efficient it is to have 10M attack instead of 1M, because more attack is always better, and really, you cannot modify your attack in a significant way, only slowly grind it up.

**Comment by Farigh** [comment]
If the basic implementation is not enough, feel free to open a new issue with the needed improvements
